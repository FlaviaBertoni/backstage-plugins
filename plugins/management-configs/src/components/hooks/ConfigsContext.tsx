import React, { createContext, useState, ReactNode, useContext } from 'react';
import useAsync from 'react-use/lib/useAsync';
import { useApi } from '@backstage/core-plugin-api';
import { ResponseError } from '@backstage/errors';
import { discoveryApiRef, fetchApiRef } from '@backstage/core-plugin-api';

export enum ConfigType {
  Secret = 'secret',
  Env = 'env',
  FeatureFlag = 'featureflag',
}

export type Config = {
  key: string;
  createdOn?: string;
  updatedOn: string;
  type?: ConfigType;
  value?: string | boolean;
  columnStatus?: React.JSX.Element;
};

interface ConfigsContextProps {
  configs: Config[];
  name: string;
  namespace: string;
  loading: boolean;
  error: Error | undefined;
  type: ConfigType;
  dispatchReloadEvent: () => void;
}

interface ConfigsProviderProps {
  children: ReactNode;
  name: string;
  namespace: string;
  type: ConfigType;
}

const ConfigsContext = createContext<ConfigsContextProps>({
  configs: [],
  name: '',
  namespace: '',
  loading: false,
  error: undefined,
  type: ConfigType.Secret,
  dispatchReloadEvent: () => {}
});

export const ConfigsProvider = ({ children, name, namespace, type }: ConfigsProviderProps) => {
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<Error | undefined>();
  const [configs, setConfigs] = useState<Config[]>([]);

  const discoveryApi = useApi(discoveryApiRef);
  const fetchApi = useApi(fetchApiRef) ?? { fetch };  

  const getConfigs = async () => {
    const baseUrlData = await discoveryApi.getBaseUrl('management-configs');

    const urlData = `${baseUrlData}/${namespace}/${name}?type=${type}`;

    const responseData = await fetchApi.fetch(urlData);
    if (!responseData.ok) {
      throw await ResponseError.fromResponse(responseData);
    }

    return await responseData.json();
  };

  const initialData = useAsync(async () => {
    const data = await getConfigs();
    setConfigs(data);
    return { data };
  }, [discoveryApi, name, type]);

  const dispatchReloadEvent = () => {
    setLoading(true);
    
    getConfigs()
      .then((data) => { setConfigs(data) })
      .catch((e: Error) => { setError(e) })
      .finally(() => { setLoading(false) });
  };

  return (
    <ConfigsContext.Provider 
      value={{ 
        configs,
        name,
        namespace,
        type,
        loading: loading || initialData.loading,
        error: error || initialData.error,
        dispatchReloadEvent 
      }}
    >
      {children}
    </ConfigsContext.Provider>
  );
};

export const useConfigsContext = () => useContext(ConfigsContext);