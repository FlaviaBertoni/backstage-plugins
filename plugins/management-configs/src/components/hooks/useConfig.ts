import { useApi } from '@backstage/core-plugin-api';
import { ResponseError } from '@backstage/errors';
import { Config, useConfigsContext } from '.';
import { discoveryApiRef, fetchApiRef } from '@backstage/core-plugin-api';

export const useConfigUpdate = () => {
  const { dispatchReloadEvent, namespace, name } = useConfigsContext();
  const discoveryApi = useApi(discoveryApiRef);
  const fetchApi = useApi(fetchApiRef) ?? { fetch };

  const update = async (item: Config) => {
    const baseUrlData = await discoveryApi.getBaseUrl('management-configs');

    const { type, key, value } = item;
    const url = `${baseUrlData}/${namespace}/${name}?type=${type}`;

    const response = await fetchApi.fetch(url, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ key, value }),
    });

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    dispatchReloadEvent();
  };

  return { update };
};

export const useConfigCreate = () => {
  const { dispatchReloadEvent, namespace, name } = useConfigsContext();
  const discoveryApi = useApi(discoveryApiRef);
  const fetchApi = useApi(fetchApiRef) ?? { fetch };

  const create = async (item: Config) => {

    const baseUrlData = await discoveryApi.getBaseUrl('management-configs');

    const { type, key, value } = item;
    const url = `${baseUrlData}/${namespace}/${name}?type=${type}`;

    const response = await fetchApi.fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ key, value }),
    });

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    dispatchReloadEvent();
  };

  return { create };
};

export const useConfigDelete = () => {
  const { dispatchReloadEvent, namespace, name } = useConfigsContext();
  const discoveryApi = useApi(discoveryApiRef);
  const fetchApi = useApi(fetchApiRef) ?? { fetch };

  const deleteConfig = async (config: Config) => {

    const baseUrlData = await discoveryApi.getBaseUrl('management-configs');

    const { type, key } = config;
    const url = `${baseUrlData}/${namespace}/${name}?type=${type}&key=${key}`;

    const response = await fetchApi.fetch(url, { method: 'DELETE' });

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    dispatchReloadEvent();
  };

  return { deleteConfig };
};

export const useConfigGet = () => {
  const { namespace, name } = useConfigsContext();
  const discoveryApi = useApi(discoveryApiRef);
  const fetchApi = useApi(fetchApiRef) ?? { fetch };

  const getConfigValue = async (item: Config) => {
    const baseUrlData = await discoveryApi.getBaseUrl('management-configs');

    const { key, type } = item;
    const url = `${baseUrlData}/${namespace}/${name}/${key}/value?type=${type}`;

    const response = await fetchApi.fetch(url);

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    return response.text();
  };

  return { getConfigValue };
};