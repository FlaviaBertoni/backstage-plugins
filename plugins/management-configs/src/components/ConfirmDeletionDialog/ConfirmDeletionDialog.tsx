import React from 'react';

import Button from '@material-ui/core/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import DialogTitle from '@mui/material/DialogTitle';

import { ConfigType } from '../hooks/ConfigsContext';

export type ConfirmDeletionDialogProps = {
    open: boolean;
    handleClose: () => void;
    handleClick: () => void;
    configKey: string;
    type?: ConfigType;
  };

export const ConfirmDeletionDialog = (props: ConfirmDeletionDialogProps) => {
    const { open, handleClose, handleClick, type, configKey } = props;
  
    return (
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="confirm-deletion-title"
      >
        <DialogTitle id="confirm-deletion-title">
          Confirm deletion
        </DialogTitle>
        <DialogContent>
          <DialogContentText>
            Do you really want to delete {type} {configKey}?
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button autoFocus onClick={handleClose}>
            Cancel
          </Button>
          <Button onClick={handleClick} autoFocus>
            Delete
          </Button>
        </DialogActions>
      </Dialog>
    );
};