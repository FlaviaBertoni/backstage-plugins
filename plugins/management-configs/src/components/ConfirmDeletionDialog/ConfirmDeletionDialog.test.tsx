import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ConfirmDeletionDialog, ConfirmDeletionDialogProps } from './ConfirmDeletionDialog';

import { ConfigType } from '../hooks/ConfigsContext';

describe('ConfirmDeletionDialog Component', () => {

    const props = {
      open: true,
      handleClose: jest.fn(),
      handleClick: jest.fn(),
      configKey: 'ft-user-forgot-password',
      type: ConfigType.FeatureFlag,
    } as unknown as ConfirmDeletionDialogProps;

  it('should render component', async () => {
    const { findByText } = render(<ConfirmDeletionDialog {...props}/>);

    await findByText(/Do you really want to delete featureflag ft-user-forgot-password/i);

    expect(document.body).toMatchSnapshot();
  });

  it('should call handleClick function when "Delete" button is clicked', () => {
    const { getByText } = render(<ConfirmDeletionDialog {...props} />);

    fireEvent.click(getByText('Delete'));

    expect(props.handleClick).toHaveBeenCalled();
  });

  it('should call handleClose function when "Cancel" button is clicked', () => {
    const { getByText } = render(<ConfirmDeletionDialog {...props} />);

    fireEvent.click(getByText('Cancel'));

    expect(props.handleClose).toHaveBeenCalled();
  });
});
