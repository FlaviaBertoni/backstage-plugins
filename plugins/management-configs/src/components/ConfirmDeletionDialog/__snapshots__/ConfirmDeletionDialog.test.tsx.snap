// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDeletionDialog Component should render component 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="confirm-deletion-title"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id="confirm-deletion-title"
        >
          Confirm deletion
        </h2>
        <div
          class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
        >
          <p
            class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiDialogContentText-root css-qfso29-MuiTypography-root-MuiDialogContentText-root"
          >
            Do you really want to delete 
            featureflag
             
            ft-user-forgot-password
            ?
          </p>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-focusVisible Mui-focusVisible"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Cancel
            </span>
            <span
              class="MuiTouchRipple-root"
            >
              <span
                class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                />
              </span>
            </span>
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-focusVisible Mui-focusVisible"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Delete
            </span>
            <span
              class="MuiTouchRipple-root"
            >
              <span
                class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                />
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
